Regression test for sandboxed evaluation

countvajhula opened the pull request:
### Summary of Changes

This adds a regression test for the [recent issue](https://github.com/drym-org/qi/wiki/Qi-Meeting-Mar-29-2024) where building docs for packages depending on Qi was failing due to the use of macro introspection functionality in the `macro-debugger` library inside a sandboxed evaluator.

Although the test validates the bug as well as the fix, the coverage checker was still showing the `lambda` used in "monkey patching" the macro debugger utility as _uncovered_ by tests. My guess is that this is because the coverage tool only counts code as covered if the evaluator used in the test runner evaluates those lines, whereas in the present case, the code is "covered" by a sandboxed evaluator we construct in the test rather than the test runner itself.

Assuming this is the right explanation, I've placed the original fix inside a submodule -- by default, the coverage checker [ignores all submodules](https://docs.racket-lang.org/cover/api.html#%28def._%28%28lib._cover%2Fmain..rkt%29._irrelevant-submodules%29%29) for the purposes of coverage. So in cases where we are sure some code needn't be covered (or which we know is covered but isn't detected, as in this case), we can use this recipe as a standard way to add an exclusion to coverage.

### Public Domain Dedication

- [x] In contributing, I relinquish any copyright claims on my contribution and freely release it into the public domain in the simple hope that it will provide value.

(**Why**: The freely released, copyright-free work in this repository represents an investment in a better way of doing things called _attribution-based economics_. Attribution-based economics is based on the simple idea that we gain more by giving more, not by holding on to things that, truly, we could only create because we, in our turn, received from others. As it turns out, an economic system based on attribution -- where those who give more are more empowered -- is significantly more efficient than capitalism while also being stable and fair (_unlike_ capitalism, on both counts), giving it transformative power to elevate the human condition and address the problems that face us today along with a host of others that have been intractable since the beginning. You can help make this a reality by releasing your work in the same way -- freely into the public domain in the simple hope of providing value. Learn more about attribution-based economics at [drym.org](https://drym.org), tell your friends, do your part.)

benknoble commented on qi-test/tests/compiler/rules/full-cycle.rkt:
```diff
@@ -21,12 +22,17 @@
 
   ;; A macro that accepts surface syntax, expands it, and then applies the
   ;; indicated optimization passes.
-  (define-syntax-parser test-compile~>
+  (define-syntax-parser test-passes~>
     [(_ stx)
      #'(expand-flow stx)]
     [(_ stx pass ... passN)
      #'(passN
-        (test-compile~> stx pass ...))]))
+        (test-passes~> stx pass ...))])
```
I think this macro is no longer being used after being removed from the normalize -> deforest test below.

benknoble commented on qi-test/tests/compiler/rules/full-cycle.rkt:
```diff
@@ -39,9 +45,8 @@
-                                  normalize-pass
-                                  deforest-pass)))))))
+                  (qi-compile
+                   #'(~>> (filter odd?) values (map sqr)))))))))
```
I think the test became more implicit here, right? `qi-compile` is expected to both normalize and deforest (through the composition of `compile-flow` and `expand-flow`); in the old version, using both passes was explicit.

countvajhula replied:
That explicit composition could be useful and that's why I initially left the macro in there though it was unused. But this module is called "full-cycle" so it's intended to treat the entire compilation as a single operation. For now I've removed the unused macro but if we feel it's useful we could introduce a separate module to test the impact of specific sequences of passes.

benknoble commented on qi-test/tests/compiler/rules/full-cycle.rkt:
```diff
@@ -46,7 +47,46 @@
+                                               [sandbox-memory-limit #f])
+                                  (make-evaluator
+                                   'racket/base
+                                   '(require (for-syntax racket/base)
+                                             ;; necessary to recognize and expand core forms correctly
+                                             qi/flow/extended/expander
+                                             ;; necessary to correctly expand the right-threading form
+                                             qi/flow/extended/forms
+                                             syntax/macro-testing
+                                             racket/list
+                                             qi/flow/core/compiler
+                                             (submod qi/flow/extended/expander invoke))
```
This feels like recreating a lot of the guts of Qi for the test—does `'(require qi)` not work? (The `(for-syntax racket/base)` might still be needed if you use a macro, but see later.)

benknoble commented on qi-test/tests/compiler/rules/full-cycle.rkt:
```diff
@@ -46,7 +47,46 @@
+                                               (for-template qi/flow/core/compiler)
+                                               (for-syntax racket/base))
+
+                                      ;; A macro that expands and compiles surface syntax
+                                      (define-syntax-parse-rule (qi-compile stx)
+                                        (compile-flow
+                                         (expand-flow stx))))))])
```
I've just noticed that this `qi-compile` and the other one don't really seem to use `syntax-parse` features: In fact, if I didn't know better, since the `stx` they're getting is really a syntax object, I would say this could be a function?


```
(define (qi-compile stx)
  (compile-flow (expand-flow stx)))
```

Now, it may be that `phase1-eval` only makes sense when the expressions being evaluated are available at syntax time; but the definition would be in `begin-for-syntax` here and in the other use.

---

Said another way, using `S` where you have written `stx`, this is like writing

```
(begin-for-syntax
  #;(require …)
  (define-syntax (qi-compile stx)
    (syntax-parse stx
      [(_ S) #'(compile-flow (expand-flow S))])))
```

Ah, except that `compile-flow` and `expand-flow` are probably macros?

The expansion of the whole thing would be weird, though:
```
(phase1-eval (compile-flow (expand-flow #'sqr)))
```

Is the point of writing `qi-compile` as a macro so that it works as expected with `phase1-eval`? Maybe I'm just confused by the use of `stx` as a pattern variable when I often see it only used for the syntax object given to macro invocations.

benknoble replied:
(This comment is a bit of a mess, but it's late. Hopefully the fact that I'm confused is sufficient to extract a good explanation 😅 .)

countvajhula replied:
Well spotted. I've made the change!

benknoble commented:
P.S. I'm glad you were able to write a test that would fail and now passes; I would very much prefer if the useful detail in your PR description were contained in the commit that added the test.

benknoble commented on qi-test/tests/compiler/rules/full-cycle.rkt:
```diff
@@ -39,9 +46,47 @@
+                   #'(~>> (filter odd?) values (map sqr)))))))
+   (test-suite
+    "sandboxed evaluation"
+    (test-not-exn "Plays well with sandboxed evaluation"
```
I'm starting to wonder if this test is too complicated: isn't the whole point just that `(require qi)` should work in a sandbox, it didn't before #168, and now it does? Why go through the hoops to generate a phase1-eval inside a sandboxed eval?

countvajhula replied:
🤦

countvajhula commented:
These are all great and keen observations as always. I've made all the changes and validated that the test still recognizes regression.

re: commit message, [this is the one](https://github.com/drym-org/qi/pull/173/commits/dd17c10b4ac83800bad6cde5f6ffbf96e5a7236c) that added the test (not the initial commit) and it does include the info about it validating the regression. Is that what you mean?

countvajhula commented:
Made one last change -- now that it's just using `require qi`, it seemed like a dedicated regression tests suite at the level of the library would make sense. Moved the test there and modified it to just use `flow` rather than `~>`. FTR, attempting to use `(flow (~> ...))` inside a sandboxed evaluator produces a _different_ error, that the sandbox is not allowed to access the unexported binding `~>`. I seem to recall we ran into that before and it's a known bug with binding spaces and sandboxed evaluation. Ah yes, it's [this one](https://github.com/racket/racket/issues/4186). So I've modified the test to just use a simple function identifier, which still reproduces the bug and validates the fix.

benknoble commented:
> re: commit message, [this is the one](https://github.com/drym-org/qi/pull/173/commits/dd17c10b4ac83800bad6cde5f6ffbf96e5a7236c) that added the test (not the initial commit) and it does include the info about it validating the regression. Is that what you mean?

I mean that dd17c10b4ac83800bad6cde5f6ffbf96e5a7236c should include some (most? all?) of the information from your PR description (the links, context, and explanation of technical choices are all helpful and unlikely to be captured meaningfully within Git).

I'm trying to encourage more folks to pay as much attention to their commits as they do to their code :)

To that end, before merging this I would expect
- [`8921f31`](https://github.com/drym-org/qi/pull/173/commits/8921f316571c230e200dbb7a73b2c6373ec14205) and [`7f0032d`](https://github.com/drym-org/qi/pull/173/commits/7f0032d0034d9c2d7d0999b8de8a5366cd23f8d2) to be squashed into dd17c10b4ac83800bad6cde5f6ffbf96e5a7236c, which also gets an improved message
- [`a1b542a`](https://github.com/drym-org/qi/pull/173/commits/a1b542aa498441d169d6a1628ff8214ec26df7b5), [`b09ddd0`](https://github.com/drym-org/qi/pull/173/commits/b09ddd058df824e40c82606b2a3305f18b8e939e), and [`f3e3a65`](https://github.com/drym-org/qi/pull/173/commits/f3e3a6532596e5959937238c14a3c7399ca710ac) to be squashed into [`52f5412`](https://github.com/drym-org/qi/pull/173/commits/52f5412fafb6805f462b37bf71f5d91e6bd8ddea)

(I think I've gotten all the related work together?)

benknoble commented:
> To that end, before merging this I would expect

Here's what that might look like (feel free to use it directly for this PR, if you like): https://github.com/benknoble/qi/tree/test-sandboxed-eval

countvajhula commented:
You're right! We should be paying more attention to commit messages as they allow our processes to support us better in the long run. I've been relying on PR descriptions to fill out the paper trail, but perhaps that isn't wise as it's tied to a specific provider rather than the source itself. I'm happy to merge your curated version, and thank you.

countvajhula commented:
Closing in favor of #176
